### PR TITLE
World rendering improvements

### PR DESCRIPTION
**Pull request by zardoy**
Basically I picked almost all changes from my fork here. I've also dramatically changed the build pipeline (migrated everything to esbuild), but I didn't pick these changes here. So it won't be in sync.
I've separated changes by commits so I do hope it is still not hard to review :)

Added missing texture from invsprite from prismarine web client. Though it would be cool to think of a way to bring old behavior for some blocks if user really wants it for some reason.

Resourcepack support is done by supplying a custom atlas image generated with [script like this](https://github.com/zardoy/prismarine-web-client/blob/b6d236d430eb86cc38aac36eb854985ddbb6a2a2/src/texturePack.ts#L164). Of course would be better to think of exposing such functionality here, but it is not my priority for now. Also added a way to supply custom block state json so you can have any texture size (e.g. 1024x)

Also in prismarine web client I'm currently going to load textures and block states before joining the server or implement a loader indicator so user have understanding that textures are being loaded.

### CODE REVIEW

**Comment by rom1504** [comment]
hi, sorry for delay

CI is failing, could you make it work?

**Comment by rom1504** [comment]
I fixed the ci

File: viewer/lib/utils.js [hunk 1]
```diff
@@ -10,7 +10,12 @@ const THREE = require('three')
 const path = require('path')
 
 const textureCache = {}
+// todo not ideal, export different functions for browser and node
 function loadTexture (texture, cb) {
+  if (process.platform === 'browser') {
+    return require('./utils.web').loadTexture(texture, cb)
```
Comment (rom1504) [inline viewer/lib/utils.js:16]: why is this needed? didn't it work before?

File: viewer/lib/utils.js [hunk 2]
```diff
@@ -22,6 +27,9 @@ function loadTexture (texture, cb) {
 }
 
 function loadJSON (json, cb) {
+  if (process.platform === 'browser') {
+    return require('./utils.web').loadJSON(json, cb)
```
Comment (rom1504) [inline viewer/lib/utils.js:31]: why is this needed? didn't it work before?

File: viewer/lib/worldrenderer.js [hunk 3]
```diff
@@ -60,23 +64,43 @@ class WorldRenderer {
...
+  updateTexturesData () {
+    loadTexture(this.texturesDataUrl || `textures/${this.version}.png`, texture => {
       texture.magFilter = THREE.NearestFilter
       texture.minFilter = THREE.NearestFilter
       texture.flipY = false
       this.material.map = texture
     })
 
-    loadJSON(`blocksStates/${version}.json`, blockStates => {
+    const loadBlockStates = () => {
```
Comment (rom1504) [inline viewer/lib/worldrenderer.js:97]: why is this needed? didn't it work before?

**Comment by rom1504** [comment]
Most things LGTM

I left 2 comments

File: viewer/lib/utils.js (hunk 2 above)
Comment (zardoy) [reply viewer/lib/utils.js:31]: As said in the body I moved to esbuild in my p viewer setup so I guess that's why I added it. If this is the only thing that blocks merging the pr and you don't want to see it you can remove it (feel free to do it since you should have push access)

File: viewer/lib/worldrenderer.js (hunk 3 above)
Comment (zardoy) [reply viewer/lib/worldrenderer.js:97]: I don't understand the phrasing of the question "didn't it work before?" - it's just an extension of the previous logic. Still works, but now you can override logic of getting block states via exposed property so you can skip network request. Ideally it should also set to this property if wasn't set before after request, fixed in my version but not here but that's not a big issue imo.

**Comment by zardoy** [comment]
Thanks!
